### ISSUE TITLE
[Postgres] int4 type conversion failure on partition compare

### ISSUE THREAD

**Issue by wseaton** [sfu-db/connector-x#149]
Related to #107, I can get queries to run now via the postgres bridge of my database (via the `cursor` protocol) but I'm running into a subtle type issue. Was wondering if you guys had any idea the cause, happy to help contribute a fix. 

```
thread '<unnamed>' panicked at 'error retrieving column 0: error deserializing column 0: cannot convert between the Rust type `i64` and the Postgres type `int4`', /github/home/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-postgres-0.7.3/src/row.rs:151:25
```

From what I can tell in the transport mapping, this conversion shouldn't be happening (https://github.com/sfu-db/connector-x/blob/main/connectorx/src/sources/postgres/typesystem.rs#L41). 

The error only happens on the `COUNT(*)` queries used to estimate the partition size. 

Example:
```
[2021-11-02T13:44:35Z DEBUG connectorx::sql] Transformed count query: SELECT count(*) FROM (SELECT cn, account, created_fiscal_yr FROM mydb.mytable WHERE createddate >= '2017-01-01') AS CXTMPTAB_PART WHERE 2019 <= CXTMPTAB_PART.created_fiscal_yr AND CXTMPTAB_PART.created_fiscal_yr < 2022
```

Maybe there is a code path there were the arrow transport is not used? Thanks!

**Comment by wangxiaoying** [comment]
Hi @wseaton , thanks for the report!

There are two places that we will directly parse the result without going through type transport.

One is the count query here: https://github.com/sfu-db/connector-x/blob/main/connectorx/src/sources/postgres/mod.rs#L308. It happens in both partition and no partition cases.

Another place is when we partition the query, we need to get the range (min and max) of the partition column first: https://github.com/sfu-db/connector-x/blob/main/connectorx-python/src/source_router.rs#L113. This will only happen when `partition_on` and `partition_num` is set.

Both of these two places happens before "Start writing" log.  Can you help to check whether the error is caused by the two cases above or after "Start writing" (where the later conversion will be done according to the typesystem you mentioned)?

**Comment by wseaton** [comment]
@wangxiaoying thanks for the quick response! The bug definitely happens after `Prepare partitions` and before `Allocate destination memory`, and I'm hitting it while both providing and not providing a partition column to split on

**Comment by wangxiaoying** [comment]
@wseaton thanks for the info. Then it is very likely that the error is caused by the count query, in which we directly use i64 to receive the result according to postgres document (count ( * ) → bigint) in https://www.postgresql.org/docs/current/functions-aggregate.html, but might not be the case in your scenario (result of count(*) is int (i4) instead of big int (i8)).

Can you help to check whether it is the case? If this is true, I think a quick fix could be match the type of the result first and then get the count (similar to get partition range).

**Comment by wseaton** [comment]
@wangxiaoying great! I think I see exactly what needs to be done, I can make the changes locally and test directly on my source system of interest to verify.

**Comment by wseaton** [comment]
@wangxiaoying any chance of a `0.2.2` alpha/prerelease to get this onto pypi?

**Comment by wangxiaoying** [comment]
> @wangxiaoying any chance of a `0.2.2` alpha/prerelease to get this onto pypi?

Yeah. It is going through our release process now, will let you know when it finishes.

**Comment by wangxiaoying** [comment]
@wseaton The new alpha version is released! `pip install -U connectorx==0.2.2-alpha.1`
